File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/ItemRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 30/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000R\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0000\n\u0002\u0010\u000b\n\u0002\b\u0003\n\u0002\u0010\b\n\u0002\b\u0002\n\u0002\u0010 \n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0002\b\u0002\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ-\u0010\t\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u000b0\n2\u0006\u0010\r\u001a\u00020\u000e2\u0006\u0010\u000f\u001a\u00020\u0010H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0011J5\u0010\u0012\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u000b0\n2\u0006\u0010\u0013\u001a\u00020\u00142\u0006\u0010\r\u001a\u00020\u000e2\u0006\u0010\u000f\u001a\u00020\u0010H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0015J\u001c\u0010\u0016\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u00170\n2\u0006\u0010\u0018\u001a\u00020\u0014H\u0016J\u001c\u0010\u0019\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u00170\n2\u0006\u0010\u001a\u001a\u00020\u0014H\u0016J\"\u0010\u001b\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u00170\u001c0\n2\u0006\u0010\u001d\u001a\u00020\u0014H\u0016R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u0082\u0002\u0004\n\u0002\b\u0019\u00a8\u0006\u001e"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/ItemRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/ItemRepository;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "itemDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/ItemDao;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;Lcom/bits/bee/bpmc/data/data_source/local/dao/ItemDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getActiveItemList", "Lkotlinx/coroutines/flow/Flow;", "Landroidx/paging/PagingData;", "Lcom/bits/bee/bpmc/domain/model/Item;", "query", "", "usePid", "", "(Ljava/lang/String;ZLkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getActiveItemListByItemGrp", "itemGrpId", "", "(ILjava/lang/String;ZLkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getItemBySelection", "", "selectionId", "getItemByVariant", "variantId", "getLastesItemList", "Lcom/bits/bee/bpmc/utils/Resource;", "page", "app_debug"})
public final class ItemRepositoryImpl implements com.bits.bee.bpmc.domain.repository.ItemRepository {
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.ItemDao itemDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    
    @javax.inject.Inject()
    public ItemRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.ItemDao itemDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Item>>> getLastesItemList(int page) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getActiveItemListByItemGrp(int itemGrpId, @org.jetbrains.annotations.NotNull()
    java.lang.String query, boolean usePid, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<androidx.paging.PagingData<com.bits.bee.bpmc.domain.model.Item>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getActiveItemList(@org.jetbrains.annotations.NotNull()
    java.lang.String query, boolean usePid, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<androidx.paging.PagingData<com.bits.bee.bpmc.domain.model.Item>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<java.util.List<com.bits.bee.bpmc.domain.model.Item>> getItemBySelection(int selectionId) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<java.util.List<com.bits.bee.bpmc.domain.model.Item>> getItemByVariant(int variantId) {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/ItemRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.map
import com.bits.bee.bpmc.data.data_source.local.dao.ItemDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.ItemImageResponse
import com.bits.bee.bpmc.data.data_source.remote.response.ItemResponse
import com.bits.bee.bpmc.domain.mapper.ItemDataMapper
import com.bits.bee.bpmc.domain.model.Item
import com.bits.bee.bpmc.domain.repository.ItemRepository
import com.bits.bee.bpmc.utils.*
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.withContext
import javax.inject.Inject

/**
 * Created by aldi on 30/03/22.
 */
class ItemRepositoryImpl @Inject constructor(
    private val apiUtils: ApiUtils,
    private val itemDao: ItemDao,
    private val defaultDispatcher : CoroutineDispatcher
) : ItemRepository {

    override fun getLastesItemList(page: Int): Flow<Resource<List<Item>>> {
        return object : NetworkDatabaseBoundResource<List<Item>, ItemResponse>(){
            override suspend fun loadFormDB(): List<Item> {
                return itemDao.getItemList().map { ItemDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<Item>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<ItemResponse>> {
                return apiUtils.getItemApiService().getInitialItem(1)
            }

            override suspend fun saveCallResult(data: ItemResponse) {
                itemDao.insertBulk(data.data.item.map { ItemDataMapper.fromNetworkToDb(it) })
            }

        }.getAsFlow()
    }

    override fun getLastesImageItem(page: Int): Flow<Resource<List<Item>>> {
        return object : NetworkDatabaseBoundResource<List<Item>, ItemImageResponse>(){
            override suspend fun loadFormDB(): List<Item> {
                return itemDao.getItemList().map { ItemDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<Item>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<ItemImageResponse>> {
                return apiUtils.getItemApiService().getImageItem()
            }

            override suspend fun saveCallResult(data: ItemImageResponse) {
                data.data.forEach { data ->
                    val item = itemDao.getItemById(data.refId)
                    item?.let {
                        itemDao.update(item.copy(
                            bucket = data.bucket,
                            objKey = data.objkey,
                            tempUrl = data.objkey.replace('/', '_')
                        ))
                    }
                }
            }

        }.getAsFlow()
    }

    override suspend fun getActiveItemListPagedByItemGrp(itemGrpId: Int, query: String, usePid : Boolean): Flow<PagingData<Item>> {
        val pagingSource = {
            if(usePid)
                itemDao.getItemByItemGrpPagedListPid(itemGrpId, query)
            else
                itemDao.getItemByItemGrpPagedList(itemGrpId, query, false)
        }

        return Pager(
            config = PagingConfig(
                pageSize = BPMConstants.BPM_LIMIT_PAGINATION,
                maxSize = BPMConstants.BPM_MAX_PAGINATION,
                initialLoadSize = BPMConstants.BPM_LIMIT_PAGINATION,
                enablePlaceholders = true
            ),
            pagingSourceFactory = pagingSource
        ).flow.mapLatest {
            it.map { ItemDataMapper.fromDbToDomain(it) }
        }.flowOn(defaultDispatcher)
    }

    override suspend fun getActiveItemListPaged(query : String, usePid : Boolean): Flow<PagingData<Item>> = Pager(
        config = PagingConfig(
            pageSize = BPMConstants.BPM_LIMIT_PAGINATION,
            maxSize = BPMConstants.BPM_MAX_PAGINATION,
            enablePlaceholders = false
        ),
        pagingSourceFactory = {
            if(!usePid)
                itemDao.getActiveItemPagedList(query, usePid)
            else
                itemDao.getActiveItemPagedListPid(query)
        }
    ).flow.mapLatest {
        it.map { ItemDataMapper.fromDbToDomain(it) }
    }.flowOn(defaultDispatcher)

    override suspend fun getFavoriteItemListPaged(query: String): Flow<PagingData<Item>> = Pager(
        config = PagingConfig(
            pageSize = BPMConstants.BPM_LIMIT_PAGINATION,
            maxSize = BPMConstants.BPM_MAX_PAGINATION,
            initialLoadSize = BPMConstants.BPM_LIMIT_PAGINATION,
            enablePlaceholders = true
        ),
        pagingSourceFactory = {
            itemDao.getFavoriteItemPagedList(query)
        }
    ).flow.mapLatest {
        it.map { ItemDataMapper.fromDbToDomain(it) }
    }.flowOn(defaultDispatcher)

    override fun getItemBySelection(selectionId: Int): Flow<List<Item>> = flow{
        emit(itemDao.getItemBySelection(selectionId).map { ItemDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)

    override fun getItemByVariant(variantId: Int): Flow<List<Item>> = flow{
        emit(itemDao.getItemByVariant(variantId).map { ItemDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)

    override fun getItemById(id: Int): Flow<Item?> {
        return flow {
            val data = itemDao.getItemById(id)
            emit(data?.let { ItemDataMapper.fromDbToDomain(data) })
        }.flowOn(defaultDispatcher)
    }

    override fun getActiveItemList(): Flow<List<Item>> = flow {
        emit(itemDao.getActiveItemList().map { ItemDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)

    override fun getActiveItemListByItemGrp(itemGrpId: Int): Flow<List<Item>> = flow {
        emit(itemDao.getActiveItemListByItemGrp(itemGrpId).map { ItemDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)

    override fun getItemByPrinterKitchen(id: Int): Flow<List<Item>> = flow {
        emit(itemDao.getItemByPrinterKitchen(id).map { ItemDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)

    override suspend fun addUpdateItem(item: Item) {
        withContext(defaultDispatcher){
            itemDao.insertSingle(ItemDataMapper.fromDomainToDb(item))
        }
    }

    override fun cariItems(query: String): Flow<List<Item>> {
        return flow {
            val data = itemDao.cariItems(query).map { ItemDataMapper.fromDbToDomain(it) }
            emit(data)
        }.flowOn(defaultDispatcher)
    }

}